Code Insights: Add search context filter v1

vovakulikov opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/33830
Closes https://github.com/sourcegraph/sourcegraph/issues/32963

## Background 

This PR updates our dilldown panel and adds a new control for the search context filter. It also fixes a few visual problems that we've been noticing recently about drilldown panel.

| Before | After |
| -- | -- |
| <img width="939" alt="Screenshot 2022-04-28 at 16 20 20" src="https://user-images.githubusercontent.com/18492575/165710802-c7166510-72c8-4c0b-98e8-299db9f8ccae.png"> | <img width="939" alt="Screenshot 2022-04-28 at 16 20 05" src="https://user-images.githubusercontent.com/18492575/165710918-d7fae7c5-8e30-44d6-9b31-919fdfd99b82.png"> |

## Todo 
- [x] Implement UI and the logic around context filter
- [x] Pass design review 
- [x] Pass tech review 
- [ ] Update our integration test in order to test search context filter [Seperate PR]

## Test plan

- Make sure that all integration tests work properly with 
- Make sure all visual details look correct according to the search UI [designs](https://www.figma.com/file/Ml5dtV9ni4mfffbfOfF9jg/Allow-search-contexts-as-filters-to-code-insights-(v0-%2333249-%26-v.1-%2333830)-%5Bready-for-dev%5D?node-id=1313%3A24762) 

<!--
  As part of SOC2/GN-104 and SOC2/GN-105 requirements, all pull requests are REQUIRED to
  provide a "test plan". A test plan is a loose explanation of what you have done or
  implemented to test this, or why this change does not need testing, as outlined in our
  Testing principles and guidelines:
  https://docs.sourcegraph.com/dev/background-information/testing_principles
  Write your test plan here after the "## Test plan" header.
-->



## App preview:

- [Web](https://sg-web-vk-add-search-context-filter-v1.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-wranezedrg.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

AlicjaSuska reviewed (changes requested):
Thank you for sharing your work @vovakulikov. Great progress!

I've created a [Design QA page](https://www.figma.com/file/Ml5dtV9ni4mfffbfOfF9jg/?node-id=1526%3A27257) with all the issues I found as well as the loom video for searching and focus issues.

vovakulikov commented:
@AlicjaSuska thanks for your review, I have fixed almost all problems but just to make sure that I haven't missed anything important I list them below

- **search issues**, as we tackled this yesterday it turned out that this is BE issue and we have the same problem. I created an issue for this problem for the team search product https://github.com/sourcegraph/sourcegraph/issues/34686
- **list styling**, I adjusted visual styles for hover and keyboard picked items accordingly to your designs, thanks for catching this. 
- **success focus states** I removed the success state from the filter field as we discussed this yesterday on sync. I left the error state actually for a regular expression and search context field (only state without error message below) because it makes sense to show that we didn't find a search context or regular expression filters have a problem. 
- **monospace** it turned out that we didn't have a proper CSS variable for it, I added monospace for all needed elements according to designs. 
- **Add color to ‘@’ to align with contexts selector** this is actually it turned out complex thing to do. I added syntax highlight for repo and -repo values though. but for context for some reason, our logic about highlighting doesn't work. I suggest moving forward without it and implementing this later.
- **Update insight only after selecting context, not when searching** So this works in the following way now: 1. You input something in the search field. 2. We are in parallel making requests for suggestion items and also request to validate that your input context value exists and render an error state if your input value doesn't exist 3. If your input value is valid we re-run the insight data card. 

https://user-images.githubusercontent.com/18492575/165894322-d882e1b1-93ed-4840-b245-860c4aebcf31.mov

unclejustin commented on client/web/src/enterprise/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/utils.ts:
```diff
@@ -0,0 +1,34 @@
+import { Validator } from '../../../../../form/hooks/useField'
```
non-blocking: we should unit test this file

vovakulikov replied:
I created an issue for integration and unit test catch up around this new functionality around filter panel

unclejustin commented on client/web/src/enterprise/insights/components/insights-view-grid/components/backend-insight/components/drill-down-filters-panel/search-context/DrillDownSearchContextFilter.tsx:
```diff
@@ -0,0 +1,113 @@
+
+    return (
+        <ComboboxList className={styles.suggestionList}>
+            {loading ? (
+                <LoadingSpinner />
+            ) : error ? (
+                <ErrorAlert error={error} />
+            ) : data ? (
+                queryBasedContexts.length === 0 ? (
+                    <span className={styles.suggestNoDataFound}>
+                        No query-based search contexts found.{' '}
+                        <Link to="/contexts/new" rel="noreferrer noopener" target="_blank">
+                            Create search context
+                        </Link>
+                    </span>
+                ) : (
+                    queryBasedContexts.map(context => (
+                        <ComboboxOption key={context.id} value={context.spec} className={styles.suggestItem}>
+                            <TruncatedText as="small" className={styles.suggestItemName}>
+                                <ComboboxOptionText />
+                            </TruncatedText>
+
+                            <TruncatedText as="small" className={styles.suggestItemDescription}>
+                                {context.description}
+                            </TruncatedText>
+                        </ComboboxOption>
+                    ))
+                )
+            ) : null}
```
non-blocking: I think this would be a lot cleaner if we used short circuits instead of nested ternarys.

For example:
```typescript
if (loading) {
        return <LoadingSpinner />
    }

    if (error) {
        return <ErrorAlert error={error} />
    }

    if (!data) {
        return null
    }

    if (queryBasedContexts.length === 0) {
        return (
            <span className={styles.suggestNoDataFound}>
                No query-based search contexts found.{' '}
                <Link to="/contexts/new" rel="noreferrer noopener" target="_blank">
                    Create search context
                </Link>
            </span>
        )
    }

    return (
        <>
            {queryBasedContexts.map(context => (
                <ComboboxOption key={context.id} value={context.spec} className={styles.suggestItem}>
                    <TruncatedText as="small" className={styles.suggestItemName}>
                        <ComboboxOptionText />
                    </TruncatedText>

                    <TruncatedText as="small" className={styles.suggestItemDescription}>
                        {context.description}
                    </TruncatedText>
                </ComboboxOption>
            ))}
        </>
    )
```

vovakulikov replied:
100% agree, I rewrote this logic with early returns, thanks!

unclejustin commented on client/web/src/enterprise/insights/components/trancated-text/TrancatedText.tsx:
```diff
@@ -1,13 +1,13 @@
-import React, { PropsWithChildren } from 'react'
+import { forwardRef } from 'react'
```
non-blocking: we should fix the typo in this file name. TrancatedText -> TruncatedText

unclejustin commented on client/web/src/integration/insights/drill-down-filters.test.ts:
```diff
@@ -76,6 +87,9 @@ describe('Backend insight drill down filters', () => {
         await driver.page.type('[name="excludeRepoRegexp"]', 'github.com/sourcegraph/sourcegraph')
 
+        // Wait until async validation of the search context field is passed
+        await delay(500)
```
non-blocking: Curious does this delay need to actually be 500ms? Would 5 work? What about 0? If it needs to be 500ms will that always be true?

vovakulikov replied:
@unclejustin this is 500 because we have an input debounce and async validation debouncing logic within the form hook. So if we remove this or set it to 0 or 5ms this test will fail.

unclejustin reviewed (approved):
LGTM @vovakulikov, just few comments.

I think in the future to keep the size down we should separate the new components from actually using them. Then we would have a nice clean separation of concerns in the PRs.

AlicjaSuska reviewed (approved):
Thank you for making changes @vovakulikov lgtm! I agree with your points in the previous comment - let's move forward with styles that are applied now :)

I've noticed one thing about 'default filters applied' message that needs correcting (probably in a separate issue):
1. open insights with default filters applied - 'Default filters applied' text is displayed, pink dot displayed
2. Change filters (reduce scope)  / reset filters - text should disappear (now, it's still being displayed), dot disappears

Here is the [loom](https://www.loom.com/share/1c21be7e23574826bc4b7ae7ec1a58e3) showing what's happening and the [Figma design](https://www.figma.com/file/Ml5dtV9ni4mfffbfOfF9jg/Allow-search-contexts-as-filters-to-code-insights-(v0-%2333249-%26-v.1-%2333830)-%5Bready-for-dev%5D?node-id=1281%3A26570) (diagram) that shows the desired flow.

vovakulikov commented:
@AlicjaSuska thanks for noticing this, I created an issue for tracking this problem with applied filter text here https://github.com/sourcegraph/sourcegraph/issues/34752

Let's continue conversation around this in that issue.
